### PR TITLE
Correct iteration order in `squire_lib::pairings::greedy`

### PR DESCRIPTION
**Pull request by zbrachinara**
This pairing algorithm was meant to process players starting from the front. However, when it failed to build a pairing using a player, it would mistakenly reject a player from the back, instead of the front. This issue is addressed by always taking off the first element, so that it is consistent across a pairing success/fail. Removal order is also changed to always be from the back ~~(and a reverse is done in the beginning to maintain order -- this can be optimized if `plyrs` is collected into a `VecDeque` by repeatedly pushing to the front)~~.

### CODE REVIEW

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:17]: I see what you're doing here, but I think it would be wise to change the arguments to this function. There is no reason this method *needs* to take in players as a `Vec`. The only thing that we care about is having ownership over the IDs. So, we can change the function like so:
```rust
pub fn greedy_pairings<I>(mut plyrs: I, ... ) -> Pairings
  where I: Iterator<Item = PlayerId>
{ ... }
```
Why would we want to do this? It helps on two fronts. First, different pairing methods currently create an iterator, apply some transformation, collect into a vector, and then pass the IDs into here. This function would be more ergonomic if we those last two steps were encapsulated in this function. It would make it easier to choose a different data structure for manipulating player ID. Second, we this would allow for an easy switch to a `VecDeque`. This type of queue allows for pushes and pops in the front and back without a ton of extra copying. This will make the `reverse` call and the later `Iterator::rev` call moot

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:29]: Not sure when this changed exactly, but the original implementation inserts zero in `index_buffer`. That does not happen here (which is fine, this is looking really good), so this should be:
```suggestion
                if id_buffer.len() == match_size - 1 {
```

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:39]: Do we need `pairing` at all now? Before, `pairing` was populated with the elements pulled out from `index_buffer`, but `retain` removes them now. I think we can simply insert `first` into `id_buffer` and push `id_buffer` into `digest`.

**Comment by TylerBloom** [comment]
Everything is looking really good so far. I would be wise to add a handful of basic unit tests. `squire_lib` is already under-tested, and with such a large refactor to this method, we should ensure that this upholds our basic assumptions.

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:33]: Very clever solution!!

**Comment by TylerBloom** [comment]
Once tests are added, I'm good to approve this after a few sanity-check tests are here. The following should suffice:
 - For the greedy pairings function (assume the match size is 4 for all of these):
   - A simple 4 person queue with no past opponents
     - Should result in a single pod of `[p1, p2, p3, p4]` with no rejections
   - A simple 5 person queue with no past opponents
     - Should result in a single pod of `[p1, p2, p3, p4]` with `p5` as the only rejection
   - A 5 person queue where players 1 and 2 have played against each other
     - Should result in a single pod of `[p1, p3, p4, p5]` with `p2` as the only rejection
   - An 8 person queue with a match size of 2 where there were prior pods of `[p1, p2]`, `[p3, p4]`, `[p5, p6]`, and `[p7, p8]`
     - Should result in a four pods of `[p1, p3]`, `[p2, p4]`, `[p5, p7]`, and `[p6, p8]`
   - An 16 person queue that is paired twice startin with no prior matches
     - First pairing should result in `[p1-4]`, `[p5-8]`, `[p9-12]`, and `[p-13-16]`
     - Second pairing should result in 4 pods. The first pod consists of the first player from each of the first pods. The second pod consists of the second players from each of the first pods, and so on.

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:214]: Can't we be more precise here? Shouldn't the first pairing contain players zero and one (and so on)?

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:243]: Similar question here. Shouldn't the first pair be players zero and two (and so on)?

File: squire_lib/src/pairings/greedy.rs
Comment (TylerBloom) [inline squire_lib/src/pairings/greedy.rs:276]: Ditto on the last two comments.

File: squire_lib/src/pairings/greedy.rs
Comment (zbrachinara) [reply squire_lib/src/pairings/greedy.rs:214]: Should the order of players in a pair also be specified?

Comment (TylerBloom) [reply squire_lib/src/pairings/greedy.rs:214]: I believe so, yes.

**Review by TylerBloom — APPROVED** [review]

**Pull request merged**
